File: sem2/hw5/task5_3/src/main/java/ru/liamin/vladimir/Calculate.java
package ru.liamin.vladimir;

/** Considers expression */
public class Calculate {

    /**
     * Considers the meaning of the expression
     * @param list list of items stored
     * @return result of the calculation of expression
     */
    public double countExpression(List list) {

        double meaning = 0;
        double variable1 = 0;
        double variable2 = 0;
        Stack stack = new Stack();

        while (list.count() > 0) {

            while (list.removeSign() == 'b') {
                stack.push(list.pop());
            }
            while ((0 < list.count()) && (list.removeSign() != 'b')) {
                variable1 = stack.pop();
                variable2 = stack.pop();
                if (list.removeSign() == '-')
                    meaning = variable2 - variable1;
                else if (list.removeSign() == '+')
                    meaning = variable2 + variable1;
                else if (list.removeSign() == '*')
                    meaning = variable2 * variable1;
                else {
                    if (variable1 == 0)
                        throw new ArithmeticException("Division by zero");
                    meaning = variable2 / variable1;
                }

                stack.push(meaning);
                list.pop();
            }
        }

        stack.clear();
        list.push(meaning, 'b', 1);
        return meaning;
    }
}

File: sem1/hw5/task5.4/calculate.cpp
#include "calculate.h"
#include "head.h"

int countExpression(char *string)
{
	Stack *stack = createStack();
	
	int meaning = 0;
	int i = 0;
	char a = '0';
	char b = '0';
	
	while (string[i] != '\n')
	{
		while ((string[i] >= '0') && (string[i] <= '9'))
		{
			add(string[i], stack);
			i++;
		}

		while ((string[i] != '\n') && ((string[i] < '0') || (string[i] > '9')))
		{
			a =  pop(stack); 
			b =  pop(stack);
			a = a - '0';
			b = b - '0';
        
			if (string[i] == '-')
				meaning = b - a;
			else if (string[i] == '+')
				meaning = b + a;
			else if (string[i] == '*')
				meaning = b * a;
			else
				meaning = int(b / a);
        	
			add(meaning + '0', stack);

			i++;
		}
	}
	
	deleteStack(stack);
	
	return meaning;
}

